oak_sign/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+// limitations under the License.
+//
+
+use anyhow::Context;
+use pem::Pem;
+use ring::{
+    rand,
+    signature::{self, Ed25519KeyPair, KeyPair},
+};
+use std::{
+    collections::HashMap,
+    fs::{read, write},
+};
+
+// PEM file tags.
+pub const PRIVATE_KEY_TAG: &str = "PRIVATE KEY";
+pub const PUBLIC_KEY_TAG: &str = "PUBLIC KEY";
+pub const SIGNATURE_TAG: &str = "SIGNATURE";
+pub const HASH_TAG: &str = "HASH";
+
+/// Convenience struct for private and public key.
+pub struct KeyPairStruct {
+    pub private_key: Vec<u8>,
+    pub public_key: Vec<u8>,
+}
+
+impl KeyPairStruct {
+    /// Generates a Ed25519 key pair.
+    pub fn generate() -> anyhow::Result<KeyPairStruct> {
+        let rng = rand::SystemRandom::new();
+        let private_key = Ed25519KeyPair::generate_pkcs8(&rng)
+            .ok()
+            .context("Couldn't generate key pair")?;
+        let key_pair = Ed25519KeyPair::from_pkcs8(private_key.as_ref())
+            .ok()
+            .context("Couldn't parse generated key pair")?;
+        let public_key = key_pair.public_key();
+        Ok(KeyPairStruct {
+            private_key: private_key.as_ref().to_vec(),
+            public_key: public_key.as_ref().to_vec(),
+        })
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Signature {
+    pub public_key: Vec<u8>,
+    pub signature: Vec<u8>,
+    pub hash: Vec<u8>,
+}
+
+impl Signature {
+    pub fn create(input: &[u8], private_key: &[u8]) -> anyhow::Result<Signature> {
+        let key_pair = signature::Ed25519KeyPair::from_pkcs8(private_key)
+            .ok()
+            .context("Couldn't parse PKCS8 encoded private key")?;
+        let hash = get_sha256(input);
+        Ok(Signature {
+            public_key: key_pair.public_key().as_ref().to_vec(),
+            signature: key_pair.sign(&hash).as_ref().to_vec(),
+            hash,
+        })
+    }
+
+    /// Verifies the signature validity.
+    pub fn verify(&self) -> anyhow::Result<()> {
+        let public_key = signature::UnparsedPublicKey::new(&signature::ED25519, &self.public_key);
+        public_key
+            .verify(&self.hash, &self.signature)
+            .ok()
+            .with_context(|| {
+                format!(
+                    "Input file signature verification failed for {}",
+                    hex::encode(&self.hash)
+                )
+            })?;
+        Ok(())
+    }
+
+    /// Parses public key, signature and SHA-256 hash encoded using PEM format.
+    /// https://tools.ietf.org/html/rfc1421
+    pub fn from_pem_file(filename: &str) -> anyhow::Result<Signature> {
+        let file =
+            read(filename).with_context(|| format!("Couldn't read signature file {}", filename))?;
+        let file_content =
+            pem::parse_many(file)
+                .iter()
+                .fold(HashMap::new(), |mut content, entry| {
+                    content.insert(entry.tag.to_string(), entry.contents.to_vec());
+                    content
+                });
```

Comment:
Couldn't this just be `.map(...).collect()`, as long as the `map` returns `(key, value)` pairs?